platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java

```diff
@@ -34,32 +39,93 @@
 import java.nio.charset.Charset;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Locale;
 import java.util.Set;
 
 public class JdkJavaCompiler implements Compiler<JavaCompileSpec>, Serializable {
     private static final Logger LOGGER = LoggerFactory.getLogger(JdkJavaCompiler.class);
 
     private final Factory<JavaCompiler> javaHomeBasedJavaCompilerFactory;
+    private final Problems problems;
 
     @Inject
-    public JdkJavaCompiler(Factory<JavaCompiler> javaHomeBasedJavaCompilerFactory) {
+    public JdkJavaCompiler(Factory<JavaCompiler> javaHomeBasedJavaCompilerFactory, Problems problems) {
         this.javaHomeBasedJavaCompilerFactory = javaHomeBasedJavaCompilerFactory;
+        this.problems = problems;
     }
 
     @Override
     public WorkResult execute(JavaCompileSpec spec) {
         LOGGER.info("Compiling with JDK Java compiler API.");
 
         ApiCompilerResult result = new ApiCompilerResult();
-        JavaCompiler.CompilationTask task = createCompileTask(spec, result);
+        DiagnosticCollector<JavaFileObject> diagnosticCollector = new DiagnosticCollector<>();
+        JavaCompiler.CompilationTask task = createCompileTask(spec, result, diagnosticCollector);
         boolean success = task.call();
+        reportProblems(diagnosticCollector);
         if (!success) {
             throw new CompilationFailedException(result);
         }
         return result;
     }
 
-    private JavaCompiler.CompilationTask createCompileTask(JavaCompileSpec spec, ApiCompilerResult result) {
+    private void reportProblems(DiagnosticCollector<JavaFileObject> diagnosticCollector) {
```

Comment:
🤔 I would move this problems logic to a separate class to separate problems logic from actual compiler logic. Ideally to some JdkJavaCompilerProblems class or maybe to some inner class.
I think that some separation would be nice, it could also help if we will do some problems logic refactorings in the future. 

What do you think about that?

Reply:
Thanks, it makes total sense, especially as said class can implement a diagnostic listener already.
Made https://github.com/gradle/gradle/blob/bhegyi/problems/java-compile-problem-reporting/platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompilerProblemListener.java as a proposed solution.